### DISCUSSION TITLE
Performance baseline

### DISCUSSION THREAD

**Discussion by visciang** [rabbitmq/ra#380] [General]
Hello!

Please bear with me, I know this is a very generic question.

I would like to ask you about the performance of the library given the design and the use cases it has been built for in the rabbitmq land.

Do you have any reference number about the library performance under some basic stable well known setup?

For example, what should I expect in term of cmd/sec (small payload, dummy "do nothing" raft state machine, no snapshotting, ...) on a stable cluster of 3 nodes with an "ideal" network (ie all nodes on the same host) ?

In the most ideal conditions "on my machine", my quick and dirty benchmark gives around 500 cmd/sec:

```
iex --name node1@127.0.0.1 -S mix
iex --name node2@127.0.0.1 -S mix
iex --name node3@127.0.0.1 -S mix

nodes = [:"node1@127.0.0.1", :"node2@127.0.0.1", :"node3@127.0.0.1"]
server_ids = for node <- nodes, do: {:quick_start, node}
for node <- nodes, do: true = Node.connect(node)
for node <- nodes, do: :rpc.call(node, :ra, :start, [])

cluster_name = :quick_start
machine = {:simple, fn command, _state -> command end, 0}
{:ok, servers_started, _servers_not_started} = :ra.start_cluster(:default, cluster_name, machine, server_ids)
leader = :ra_leaderboard.lookup_leader(cluster_name)
:timer.tc(fn -> Enum.each(1..5000, &:ra.process_command(leader, &1)) end)
```

This should be the reference baseline for a single RAFT cluster of 3 nodes.

Do I miss something? Do you have any hint or advice about the performance topic?

Thank you all!
Giovanni

PS: how do rabbitmq quorum queues leverage ra to push the message throughput?

**Comment by kjnilsson** [comment]
Your tests is testing a completely synchronous workload which means every call to `ra:process_command/2` will wait for the full fsync replication cycle before returning. Thus thoughput will depend very much on the speed of your disks.

If you use ra:pipeline_command instead in a manner similar to what is done in `ra_bench` you will see very different results.

**Reply by visciang** [reply]
Yes, I wanted to see the synchronous "safe" end-to-end path.

This to test the scenario where the clients need to know the command results and be sure it has been applied by the majority of the cluster servers (ie the case of important commands not "fire and forget" ones).

Thank you @kjnilsson

**Reply by kjnilsson** [reply]
Use `ra:pipeline_command/3+` with a correlation id and you will get an async notification including the return value from the state machine.
